### PR TITLE
Increase the default resync period in the sample controller.

### PR DESCRIPTION
**Pull request by mattmoor**
See the detailed comment inline, but this default is a bad example for a variety of reasons.

**What type of PR is this?**

/kind cleanup
/kind documentation

**What this PR does / why we need it**:

The current sample uses a resync period that is very low and doesn't explain what it is.  This explains things and increases the value to match sigs.k8s.io/controller-runtime.

**Which issue(s) this PR fixes**:

I haven't opened an issue, but I have whined on Twitter: https://twitter.com/pires_oss/status/1106882915470856192

🤦‍♂️ 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

No

### CODE REVIEW

File: staging/src/k8s.io/sample-controller/main.go [hunk 1]
```diff
@@ -58,8 +58,23 @@ func main() {
...
 	}
 
-	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, time.Second*30)
-	exampleInformerFactory := informers.NewSharedInformerFactory(exampleClient, time.Second*30)
+	// Informers trigger events on resource changes, which typically queues
+	// a reconciliation of that resource.  The resync period acts as a
+	// safety net to protect against dropped events, and the informer will
+	// signal a "nop" event to trigger a reconciliation of every resource
+	// on this interval.  Small values here have an assortment of problems:
+	//  1. It can mask unhandled events.  If an event is not handled then
```
Comment (smarterclayton) [inline staging/src/k8s.io/sample-controller/main.go]: So part of the reason historically for this is that this is a feature while iterating on a controller.  Most controllers that will ever be written are small, have a small working set, and may never get to the point of needing more.  Brute forcing sync is an acceptable solution for that problem domain.

While I’m not opposed to increasing the defaults, its probably also important to communicate the problems this comment describes by changing the parameter name to something that communicates the nature better, forcing someone to set a default, or changing the signature of the higher level code to give this context.  I’m worried that this trades one problem for another, while we want to force controller authors up front to ask “what am I doing when I set resyncninternal”.

10 hours is dangerously high from the perspective of “you won’t catch this problem until 10 hours after you’ve deployed it” as well.

Comment (smarterclayton) [reply staging/src/k8s.io/sample-controller/main.go]: As suggested on Twitter, how about 10m?  Controller runtime should probably not be at 10h either, 10m would be a nice compromise.

File: staging/src/k8s.io/sample-controller/main.go [hunk 2]
```diff
@@ -58,8 +58,23 @@ func main() {
...
+	//    the associated resources will still be reconciled quickly and
+	//    likely not caught by e2e testing.
+	//  2. Doesn't scale well.  The default rest QPS limit is 5, so assuming
```
Comment (lavalamp) [inline staging/src/k8s.io/sample-controller/main.go]: This reason shouldn't be true--well-written controllers shouldn't trigger network traffic unless a resource actually changes.

File: staging/src/k8s.io/sample-controller/main.go [hunk 3]
```diff
@@ -58,8 +58,23 @@ func main() {
 		klog.Fatalf("Error building example clientset: %s", err.Error())
 	}
 
-	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, time.Second*30)
-	exampleInformerFactory := informers.NewSharedInformerFactory(exampleClient, time.Second*30)
+	// Informers trigger events on resource changes, which typically queues
+	// a reconciliation of that resource.  The resync period acts as a
+	// safety net to protect against dropped events, and the informer will
```
Comment (lavalamp) [inline staging/src/k8s.io/sample-controller/main.go]: "Acts as a safety net" sounds like a positive thing. "Can be used to limp along in the presence of controller bugs" sounds more like how I feel, anyway. :)

This comment doesn't capture some details I think are important:

1. A legitimate usage for this is polling non-k8s resources (e.g., Load Balancers in your cloud provider). Choose the period based on responsiveness vs cloud-provider (NOT k8s) qps usage.
2. Ordinary network failures etc means that your controller COULD get a resync at any time. You must test and handle this. A small value here in tests ensures that this critical code path is exercised.
3. Consider e.g. setting a metric or logging whenever your controller takes action on a resync. Unless you're polling some other resource, it's probably a bug.

File: staging/src/k8s.io/sample-controller/main.go (hunk 1 above)
Comment (lavalamp) [reply staging/src/k8s.io/sample-controller/main.go]: 10h might as well be never--which is how I think many controllers ought to run :)

10m is a decent default.

As much as I hate flags, I think there should probably be a flag, to allow and/or encourage people to run with different settings in test vs prod.

Comment (mattmoor) [reply staging/src/k8s.io/sample-controller/main.go]: I have no skin in this game.  Paging @pwittrock :)

**Pull request closed without merging**